File: LeetCode/all-nodes-distance-k-in-binary-tree.cpp
// https://leetcode.com/problems/all-nodes-distance-k-in-binary-tree/
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution
{
public:
    vector<int> distanceK(TreeNode *root, TreeNode *target, int K)
    {
        vector<int> res;

        if (!root)
        {
            return res;
        }

        // Keeps track of the parent of each node
        unordered_map<TreeNode *, TreeNode *> parents;
        // Keeps track of which nodes have been visited
        unordered_map<TreeNode *, int> visited;

        // Populate the parents map
        buildMap(root, NULL, parents);

        queue<TreeNode *> q;

        // Start Level Order with the target node
        q.push(target);
        visited[target] = 1;

        int level = 0;

        while (!q.empty())
        {

            if (level == K)
            {
                break;
            }

            int nodeCount = q.size();
            while (nodeCount--)
            {
                auto curr = q.front();
                q.pop();

                if (curr->left && visited[curr->left] != 1)
                {
                    q.push(curr->left);
                    visited[curr->left] = 1;
                }
                if (curr->right && visited[curr->right] != 1)
                {
                    q.push(curr->right);
                    visited[curr->right] = 1;
                }
                if (parents[curr] && visited[parents[curr]] != 1)
                {
                    q.push(parents[curr]);
                    visited[parents[curr]] = 1;
                }
            }
            level++;
        }

        // We broke out of the loop when it contained the nodes for level k
        // (remember that we started at level 0, so AFTER k-1'th iteration you had the required nodes in queue)
        while (!q.empty())
        {
            res.push_back(q.front()->val);
            q.pop();
        }

        return res;
    }

    void buildMap(TreeNode *root, TreeNode *parent, unordered_map<TreeNode *, TreeNode *> &parents)
    {

        parents[root] = parent;

        if (root->left)
        {
            buildMap(root->left, root, parents);
        }

        if (root->right)
        {
            buildMap(root->right, root, parents);
        }
    }
};

File: LeetCode/all-nodes-distance-k-in-binary-tree.java
// https://leetcode.com/problems/all-nodes-distance-k-in-binary-tree/

import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode(int x) { val = x; }
 * }
 */
class Solution {
  public List<Integer> distanceK(TreeNode root, TreeNode target, int k) {
    List<Integer> result = new ArrayList<>();

    if (root == null)
      return result;

    // Keeps track of the parent of each node.
    HashMap<TreeNode, TreeNode> parents = new HashMap<>();
    // Keeps track of all the visited nodes.
    HashSet<TreeNode> visited = new HashSet<>();

    // populate the parents map.
    buildMap(root, null, parents);

    Queue<TreeNode> q = new ArrayDeque<>();
    q.offer(target);
    visited.add(target);

    int level = 0;

    // Loop to get the nodes at K'th level in the queue.
    while (!q.isEmpty()) {
      // We currently have nodes at level 'level' in the queue.
      // If level == k, break. The nodes in the queue are the result set.
      if (level == k)
        break;

      int nodesAtThisLevel = q.size();
      while (nodesAtThisLevel-- > 0) {
        TreeNode curr = q.poll();

        if (curr.left != null && !visited.contains(curr.left)) {
          q.offer(curr.left);
          visited.add(curr.left);
        }

        if (curr.right != null && !visited.contains(curr.right)) {
          q.offer(curr.right);
          visited.add(curr.right);
        }

        if (parents.get(curr) != null && !visited.contains(parents.get(curr))) {
          q.offer(parents.get(curr));
          visited.add(parents.get(curr));
        }
      }
      level++;
    }

    // We broke out of the loop when it contained the nodes for level k
    // (remember that we started at level 0, so AFTER k-1'th iteration you had the
    // required nodes in queue)
    while (!q.isEmpty()) {
      result.add(q.poll().val);
    }

    return result;
  }

  void buildMap(TreeNode node, TreeNode parent, HashMap<TreeNode, TreeNode> parents) {
    if (node == null)
      return;

    parents.put(node, parent);

    if (node.left != null)
      buildMap(node.left, node, parents);
    if (node.right != null)
      buildMap(node.right, node, parents);
  }
}